File: Session8/C++/SearchInLinkedList.cpp
#include <iostream>
using namespace std;

class Node{
public:
    int data;
    Node *next;

    Node(int d)
    {
        data = d;
        next = NULL;
    }
};

void insertionAtHead(Node*&head, int data)
{
    Node *new_Node = new Node(data);
    new_Node->next = head;
    head = new_Node;
}

bool searchIterative(Node*head, int key)
{
    while(head!=NULL)
    {
        if(head->data==key)
        {
            return true;
        }
        head = head->next;
    }
    return false;
}

bool searchRecursive(Node*head, int key)
{
    // base case
    if(head==NULL){
        return false;
    }
    if(head->data==key){
        return true;
    }
    return searchRecursive(head->next, key);
}

int main() {

    Node* head = NULL;  

    insertionAtHead(head, 1);
    insertionAtHead(head, 2);
    insertionAtHead(head, 3);
    insertionAtHead(head, 4);
    insertionAtHead(head, 5);

    if(searchIterative(head, 1)){
        cout<<"Node is Present"<<endl;
    }
    else{
        cout<<"Not Present"<<endl;
    }

    if(searchRecursive(head, 9)){
        cout<<"Node is Present"<<endl;
    }
    else{
        cout<<"Not Present"<<endl;
    }

    return 0;
}

File: Session8/Java/SearchInLinkedList.java
class SearchInLinkedList
{
	Node head; 
	class Node
	{
		int data;
		Node next;
		Node(int d) 
		{
			data = d; 
			next = null; 
		}
	}

	public void insertionAtHead(int data)
	{

		Node new_node = new Node(data);
		new_node.next = head;
		head = new_node;
	}

    boolean searchIterative(int key)
    {
        Node temp = head;
        while(temp!=null)
        {
            if(temp.data==key)
            {
                return true;
            }
            temp = temp.next;
        }
        return false;
    }

    boolean searchRecursive(Node head, int key)
    {
        // base case
        if(head==null){
            return false;
        }
        if(head.data==key){
            return true;
        }
        return searchRecursive(head.next, key);
    }

	public static void main(String[] args)
	{
		SearchInLinkedList llist = new SearchInLinkedList();

		llist.insertionAtHead(1);
		llist.insertionAtHead(2);
		llist.insertionAtHead(3);
		llist.insertionAtHead(4);
		llist.insertionAtHead(5);

        if(llist.searchIterative(10)){
            System.out.println("Node is Present");
        }
        else{
            System.out.println("Not Present");
        }

        if(llist.searchRecursive(llist.head, 1)){
            System.out.println("Node is Present");
        }
        else{
            System.out.println("Not Present");
        }
	}
}
